store/tikv/region_cache.go

```diff
@@ -45,24 +48,36 @@ func (c *RegionCache) GetRegionByVerID(id RegionVerID) *Region {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 
-	return c.regions[id]
+	if r, ok := c.regions[id]; ok {
+		return r.Clone()
+	}
+	return nil
 }
 
 // GetRegion find in cache, or get new region.
 func (c *RegionCache) GetRegion(key []byte) (*Region, error) {
-	if r := c.getRegionFromCache(key); r != nil {
+	c.mu.RLock()
+	r := c.getRegionFromCache(key)
+	c.mu.RUnlock()
+	if r != nil {
 		return r, nil
 	}
 	r, err := c.loadRegion(key)
-	return r, errors.Trace(err)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.insertRegionToCache(r)
+	return c.getRegionFromCache(key), nil
```

Comment:
@siddontang Other thread may have inserted a Region with the same ID before, we should always return the Region in cache. I'll add a note here.